### PR TITLE
feat!: Default integers to u64

### PR DESCRIPTION
**Pull request by guipublic**
# Description

Remove the deprecation warning about loop iterators.
Default integers to u64 instead of field, since it was only used for loops and array indexing.

## Problem\*

Related to #2723

## Documentation

- [X] This PR requires documentation updates when merged.

  

  - [ ] I will submit a noir-lang/docs PR.

  

  - [X] I will request for and support Dev Rel's help in documenting this PR.

  
  

## Additional Context

# PR Checklist\*

- [X] I have tested the changes locally.
- [X] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

**Review by jfecher — CHANGES_REQUESTED** [review]
Defaulting all integers to u64 instead of Field would be a very large breaking change. I'm blocking this for now not necessarily because I disagree with the change, but to get some more discussion before it goes through.

Changing the default type away from a Field seems odd to me. If I write `let x = 3 + 2;` I think I'd expect `x` to be a Field since it is the most "basic" or fundamental underlying type. Using `u64` over Field would also lead to performance loss on overflow checks I'd think.

File: compiler/noirc_frontend/src/monomorphization/mod.rs [hunk 1]
```diff
@@ -325,13 +328,15 @@ impl<'interner> Monomorphizer<'interner> {
...
-                r#type: Self::convert_type(&cast.r#type),
+                r#type: self.convert_type(&cast.r#type),
                 location: self.interner.expr_location(&expr),
             }),
 
             HirExpression::For(for_expr) => {
+                self.is_range_loop = true;
                 let start = self.expr(for_expr.start_range);
                 let end = self.expr(for_expr.end_range);
+                self.is_range_loop = false;
```
Comment (jfecher) [inline compiler/noirc_frontend/src/monomorphization/mod.rs:339]: I'm worried this check is a bit ad-hoc in that if the range expression contains other expressions, those would all be defaulted to `u64` as well. E.g. in `for _ in 0 .. { let x = y*10000000; 27 } { ... }`.

This is contrived though, I doubt there are any real programs that have any reason to do this.

If we ever add the `Integer` type variable kind as in https://github.com/noir-lang/noir/issues/1991 that would fix this more cleanly I think. Then any integer using comparisons or in the range of a for loop can be defaulted to u64 automatically.

**Review by jfecher — APPROVED** [review]
This looks fine for now. There's one contrived case I can see with this PR changing all integers to default to u64 within range expressions, but this would only be encountered by users writing somewhat questionable code.

Once we add `TypeVariableKind::Integer` as in #1991, we can remove this check entirely since any integer used in comparisons or a for loop's range would be defaulted to u64 automatically.

File: compiler/noirc_frontend/src/monomorphization/mod.rs (hunk 1 above)
Comment (kevaundray) [reply compiler/noirc_frontend/src/monomorphization/mod.rs:339]: #1991 is merged as completed, should we open up another issue for this?

**Pull request merged**